### ISSUE TITLE
[3.0.2] Regression on security.encoder_factory

### ISSUE THREAD

**Issue by havvg** [symfony/symfony#17864]
It seems there is a regression in `v3.0.2` when you require a password encoder using a factorised service.

```yaml
    user.password_encoder:
        class: Symfony\Component\Security\Core\Encoder\PasswordEncoderInterface
        factory: ['@security.encoder_factory', 'getEncoder']
        arguments:
            - 'Application\Security\User'

    user.factory:
        class: Application\Domain\User\Model\UserFactory
        arguments:
            - '@user.password_encoder'
```

```
 $ behat -o /dev/null && composer up symfony/symfony && behat -o /dev/null
Loading composer repositories with package information
Updating dependencies (including require-dev)
  - Updating symfony/symfony (v3.0.1 => v3.0.2)
    Checking out 18c3d4f356931a5b6a4afb0cc679a2c58931c795

Writing lock file
Generating autoload files
Changelogs summary:

 - symfony/symfony updated from v3.0.1 to v3.0.2
   See changes: https://github.com/symfony/symfony/compare/v3.0.1...v3.0.2
   Release notes: https://github.com/symfony/symfony/releases/tag/v3.0.2

> Incenteev\ParameterHandler\ScriptHandler::buildParameters
Updating the "app/config/environments/parameters.yml" file
> Application\Composer\BootstrapHandler::ensureDirectoriesExist
Ensuring the var directory exists.
> Sensio\Bundle\DistributionBundle\Composer\ScriptHandler::clearCache

 // Clearing the cache for the dev environment with debug true


 [OK] Cache for the "dev" environment (debug=true) was successfully cleared.


> Sensio\Bundle\DistributionBundle\Composer\ScriptHandler::installAssets

 Trying to install assets as relative symbolic links.

 --- ----------------- ------------------
      Bundle            Method / Error
 --- ----------------- ------------------
  ✔   FrameworkBundle   relative symlink
 --- ----------------- ------------------


 [OK] All assets were successfully installed.


  [Symfony\Component\DependencyInjection\Exception\ServiceNotFoundException]
  You have requested a non-existent service "security.encoder_factory.generic". Did you mean this: "security.encoder_factory"?


behat [-s|--suite SUITE] [-f|--format FORMAT] [-o|--out OUT] [--format-settings FORMAT-SETTINGS] [--init] [--lang LANG] [--name NAME] [--tags TAGS] [--role ROLE] [--story-syntax] [-d|--definitions DEFINITIONS] [--append-snippets] [--no-snippets] [--strict] [--order ORDER] [--rerun] [--stop-on-failure] [--dry-run] [--] [<paths>]
```

**Comment by xabbuh** [comment]
Can you provide some steps that are necessary to reproduce your issue on a naked Standard Edition without the need of setting up Behat?

**Comment by xabbuh** [comment]
@havvg Just noticed your referenced fork.

Status: Reviewed

**Comment by xabbuh** [comment]
This also affects Symfony 2.7 and 2.8 (it looks like this issue was introduced by #17554).

**Comment by xabbuh** [comment]
see #17866 for the fix

**Comment by xabbuh** [comment]
#17867 fixes this for the `2.3` branch
